### ISSUE TITLE
Newlines and Carriage Returns Following `$()` Command Substitution Cause Bash Parse Error

### ISSUE THREAD

**Issue by zedseven** [git-for-windows/git#3303]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing (apologies if I missed one)

### Setup

I'm Using `Git-2.32.0-64-bit`, the latest release as of writing:
```
$ git --version --build-options

git version 2.32.0.windows.1
cpu: x86_64
built from commit: 4c204998d0e156d13d81abe1d1963051b1418fc0
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

I'm running Windows 10 64-bit:
```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.19041.1052]
```

My installation options:
```
$ cat /etc/install-options.txt

Editor Option: VIM
Custom Editor Path:
Default Branch Option: main
Path Option: CmdTools
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Core
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Enabled
Enable FSMonitor: Disabled
```

#### Any other interesting things about your environment that might be related to the issue you're seeing?

I decided to make a custom `.bashrc` for my Windows installation, including setting up a custom `$PS1`. This is where I encountered the issue.

#### Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

I'm running MSYS2, Git Bash.

#### What commands did you run to trigger this issue?

Run the following:
```bash
export PS1='$(echo test)\n'
# or
export PS1='$(echo test)\r'
```
After running either of these, every time `$PS1` is evaluated, bash outputs an error:
```
bash: command substitution: line 1: syntax error near unexpected token `)'
bash: command substitution: line 1: `echo test)'
```
https://user-images.githubusercontent.com/25164338/124189132-ca1d9380-daaf-11eb-88c9-edfc43ffad88.mp4

https://user-images.githubusercontent.com/25164338/124189997-187f6200-dab1-11eb-8813-e08da95f16d9.mp4

The cause is the newline or carriage return character. If it's removed, the `$PS1` works as expected:

https://user-images.githubusercontent.com/25164338/124189285-018c4000-dab0-11eb-808a-5f4a06fc1576.mp4

This seems to be a long-known issue exclusive to Git for Windows (https://stackoverflow.com/a/41925324/6003488) however there's more that I found.

Replacing the `$()` with backticks avoids the issue for both the newline and carriage return:
```bash
export PS1='`echo test`\n'
export PS1='`echo test`\r'
```
https://user-images.githubusercontent.com/25164338/124189411-30a2b180-dab0-11eb-889c-9d5ebaf331e7.mp4

https://user-images.githubusercontent.com/25164338/124193418-4c10bb00-dab6-11eb-8bd7-72909f8b681b.mp4

And replacing the newline character with it's ASCII character code also avoids the issue, but this only works for the newline:
```bash
# https://stackoverflow.com/a/41925324/6003488
# works:
export PS1='$(echo test)\012'
# doesn't work:
export PS1='$(echo test)\015'
```
https://user-images.githubusercontent.com/25164338/124189553-6c3d7b80-dab0-11eb-9660-2c9ef4e19ba8.mp4

https://user-images.githubusercontent.com/25164338/124193246-048a2f00-dab6-11eb-804e-901977d3639c.mp4

Outputting the newline as a direct character sequence works, but doing the same with a carriage return doesn't work:
```bash
# https://stackoverflow.com/a/21561763/6003488
# works:
export PS1='$(echo test)'$'\n'
# doesn't work:
export PS1='$(echo test)'$'\r'
```
https://user-images.githubusercontent.com/25164338/124193991-4ff10d00-dab7-11eb-8733-9b07d15f3260.mp4

https://user-images.githubusercontent.com/25164338/124194017-57b0b180-dab7-11eb-8730-cd733a15ce1e.mp4

Outputting the characters from a separate function seems to work so long as the function call is done every time `$PS1` is evaluated, rather than upon definition:
```bash
# https://stackoverflow.com/a/37074809/6003488
function echocr() {
    echo -e -n "\r"
}
# works (because echocr is called every evaluation):
export PS1="\$(echo test)\$(echocr)"
# doesn't work (because echocr is evaluated once and put in $PS1 directly):
export PS1="\$(echo test)$(echocr)"
```
https://user-images.githubusercontent.com/25164338/124194494-356b6380-dab8-11eb-846b-6217882e0a3e.mp4

https://user-images.githubusercontent.com/25164338/124194498-38feea80-dab8-11eb-9d3d-fc2b4fd736bf.mp4

#### What did you expect to occur after running these commands?

I expected the new `$PS1` variable to be evaluated without error, the same way the bash string would be if used outside of `$PS1`.

#### What actually happened instead?

I got an obscure, unhelpful error that indicated the bash installation was having trouble parsing the new `$PS1` variable, even though it should be valid bash.

### Additional Notes

Apologies if there exists an issue that mentions this already. I looked and didn't find one, and figured I'd gather all the information I know about the issue and post it so it can hopefully finally be resolved.

I don't know much more about the issue, other than this. I was able to work around the problem by using the separate function method for my use of a carriage return, but obviously that's not a proper solution.

**Comment by dscho** [comment]
Does this also happen in a regular [MSYS2](https://msys2.github.io/) Bash?

**Comment by dscho** [comment]
BTW you can easily work around that issue:

```sh
PS1='$(echo test)
'
```

(Note the literal line break before the closing quote, and also note that you need not `export` this variable.)

**Comment by zedseven** [comment]
Just tested it, and it looks like it does (though I haven't tested every single case, the simple case of `PS1='$(echo test)\n'` fails).

Does this mean the issue needs to be raised somewhere else?

**Comment by zedseven** [comment]
The problem that led me to actually create this issue was that of the carriage return - enough workarounds exist for the newline character, but very few also work with the carriage return which led to a bit of a headache.

As I said before I was able to work around this, but it was a bit annoying to say the least.

**Comment by dscho** [comment]
> 
> 
> Just tested it, and it looks like it does (though I haven't tested every single case, the simple case of `PS1='$(echo test)\n'` fails).
> 
> Does this mean the issue needs to be raised somewhere else?

The next step to test would be [Cygwin](https://cygwin.com/). MSYS2 is a close derivative of Cygwin, and to pinpoint where exactly the bug exists, it would be good to know whether Cygwin shares the behavior of MSYS2's.

**Comment by zedseven** [comment]
Looks like Cygwin doesn't have an issue with any of the tests, and behaves as expected.

https://user-images.githubusercontent.com/25164338/125713964-42453d1d-ed46-4590-b656-0bfe19581cc4.mp4


I also went back and tested MSYS2 again, and found that newer installations of it still failed, but with different error messages.

An older installation (Bash 4.4.023-1):

https://user-images.githubusercontent.com/25164338/125714145-e3d9373e-c258-4f3e-943c-34a815a3713f.mp4

A newer installation (Bash 5.1.004-1):

https://user-images.githubusercontent.com/25164338/125714311-d38ea6e5-fcde-409c-af96-27ff1bbd60f6.mp4

(I just wrote the Bash versions above as markers of their approximate age, since afaik MSYS2 doesn't have a canonical version. That's also why I included the output of `pacman -Q`)
